massenergize/api
# Issue #370 - Asynchronous downloads

This PR seeks to eliminate the waiting time during data download on the platform by leveraging async task queues. When a user initiates a download, the data processing is taken out of the request-response pipeline and placed in an async queue. Thereby, releasing the pipeline for processing other tasks. The data is emailed to the admin who requested it upon completion.

**Procedure**
1. The handler receives the request and routes it to the service layer
2. The service pushed the data processing to a queue via `download_data function` and return a response immediately
3. Upon completion the data is emailed to the admin who requested it.

### HOW TO TEST

- Backend setup
The implementation uses Redis and amazon SQS as celery message broker locally and remote respectively. To setup redis,
    1. Run the command `brew install redis` in your terminal to install redis.
    2. After the installation is complete run `brew services start redis` to start redis service
    3. To see the `port` on which your redis is running, run `redis-server`
    
   
<img width="522" alt="Screenshot 2022-04-23 at 4 24 31 PM" src="https://user-images.githubusercontent.com/42780120/164915795-a94e5184-6eda-4b27-af64-c6f46141f557.png">

    
    4. Copy the `port` and add this code to local.env file 
    
    ```
    CELERY_LOCAL_REDIS_BROKER_URL='redis://localhost:PORT/0'
    ```
    

     

Now that you have setup redis navigate to massEnergize api directory and run the command below in your terminal to  install the required python packages.

```python
pip install celery
pip install redis
pip install  pycurl ==7.19.3.1
pip install 'celery[sqs]'
```
**NOTE
For linux platform, some other dependencies are required to run pycurl successfully. Run this command to install
```
sudo apt-get update --fix-missing && sudo apt-get install -y libcurl4-openssl-dev
```


After successfully installing these packages, run the commands below in separate terminals to start django and celery servers.

```makefile
1. make start
2. make celery # for local testing 
	or make celery-remote # for remote testing
```

Hopefully you have complete setup if no error is thrown in any of the terminals.

- Frontend Setup
    - Checkout `Tahiru-Async-downloads` branch on [admin frontend](https://github.com/massenergize/frontend-admin) repository
    - Run the frontend server
    - login and try downloading (users, actions or communities) on the dashboard. You should get an immediate response indicating that the data will be emailed when the data preparation is complete.
